rfcs/2021-08-06-8619-framing-and-codecs-sources.md

```diff
@@ -0,0 +1,247 @@
+stream of bytes into byte frames (chunks of data that have finite size and
+contain complete messages), while a _Codec_ converts between a byte frame and
+structured data.
+
+Conceptually, we want reusable pieces of logic which would allow us to collapse
+a component + transform into a component with codec config in places where the
+transform was merely used to convert between encoding formats, e.g. from `bytes`
+to `json` or vice versa.
+
+## Scope
+
+For this part, we focus on the source/decoding side and leave the sink/encoding
+side to a separate RFC.
+
+The scope of this RFC concerns at which level framing and decoding operates, how
+these framers and decoders can be configured, and how they can be shared in a
+uniform way. It does not cover any specific implementation for a framer or a
+decoder. Also see [future work](#future-work) for extended goals that are not
+covered by this RFC.
+
+## Motivation
+
+Currently, we have no explicit abstraction that is responsible for handling
+framing and decoding, such that each source may implement these in their own
+way. This has led to inconsistencies in available options and poses additional
+maintenance burden. Components have their own defaults, making
+behavior unpredictable and surprising users, e.g. as documented in
+[#3453](https://github.com/timberio/vector/issues/3453).
+
+## Internal Proposal
+
+To expose framing and decoding to sources in a uniform way, we specify a common
+configuration struct:
+
+```rust
+pub struct DecodingConfig {
+    framing: Option<FramingConfig>,
+    decoding: Option<ParserConfig>,
+}
+```
+
+while the framing method and parser used to deserialize into a structured event
+can be selected from an enum:
+
+```rust
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum FramingConfig {
+    NewlineDelimited,
+    CharacterDelimited {
+        delimiter: char,
+        max_length: Option<usize>,
+    },
+    OctetCounting {
+        max_length: Option<usize>,
+    },
+    ...
+}
+
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ParserConfig {
+    Bytes,
+    Json {
+        ...
+    },
+    Syslog,
+    ...
+}
+```
+
+The `DecodingConfig` exposes a `build` method to create a Vector `Decoder` which
+implements the `tokio_util::codec::Decoder<Item = (Event, usize)>` trait. That
+way we can produce `Event`s from `ByteMut` by repeated calls to
+`decode`/`decode_eof` either on a byte stream or a byte message. The additional
+`usize` item conveys how many bytes were read to produce the particular `Event`,
+so that this information can be passed along to e.g. the internal event log.
+
+Internally, the `Decoder` holds a framer and a parser:
+
+```rust
+pub struct Decoder {
+    framer: Box<dyn tokio_util::codec::Decoder<Item = Bytes> + Send + Sync>,
+    parser: Box<dyn Parser + Send + Sync>,
+}
+```
+
+with the `Parser` trait being defined as:
+
+```rust
+pub trait Parser {
+    fn parse(&self, bytes: Bytes) -> crate::Result<Event>;
```

Comment:
While implementing this, the single `Event` return type turned out to be a bit unfortunate, since we can't emit multiple events e.g. from parsing a JSON array.

In some places we are emitting information about the `byte_size` (e.g. `SocketEventReceived` or `SyslogEventReceived`)  for each individual event, therefore the decoder emits `(Event, usize)` as an item. Now I'm wondering when accounting for multiple events, should the signature change to `(Vec<Event>, usize)` or do we somehow try to find out how many bytes were read for each parsed event? For the JSON case this would likely mean that some bytes would not be accounted for (the opening and closing braces).

Reply:
Do we normally parse it into multiple events from the get-go? I thought that's why we've added the `unnest` function to VRL

Reply:
Yes, otherwise I would have also liked to defer this question.

E.g. we support `ndjson`/`json` in HTTP and parse multiple events either separated by newline or by JSON array entries

Reply:
What I have in mind: we either change the metrics to emit e.g. `*FrameReceived` and potentially push more events into the pipeline than frames reported (since one frame can contain multiple events), or we remove the `byte_size` from the `*EventReceived` metrics

Reply:
I think we can decouple the bytes metrics from the event metrics. Framers can emit byte counts and parsers can emit event counts.

For parsing multiple events from a single frame, we could potentially change the parse method to something similar to transforms:

```rust
fn parse(&self, bytes: Bytes, output: &mut Vec<Event>) -> crate::Result<()>;
```

If we still want to keep single-event parsers simple, we could use the above method as a broader `MultiParser` trait (with a better name, hopefully), and then:

```rust
impl<T> MultiParser for T where T: Parser {
    fn parse(&self, bytes: Bytes, output: &mut Vec<Event>) -> crate::Result<()>; {
        let event = Parser::parse(self, input)?;
        output.push(event);
        Ok(())
    }
}
```

Might not be worth two traits though.

Reply:
I just tried implementing this such that our `Decoder` has `Item = (Vec<Event>, usize)`.

This has some unfortunate consequences:
- the `Decoder` `Item` requires allocating a `Vec` since traits do not support associated lifetimes yet (see https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md). Otherwise at least returning a reference to a `Vec` that we reuse would have been preferable
- due to the above, the optimization for `(..., output: &mut Vec<Event>) -> crate::Result<()>` over `-> crate::Result<Vec<Event>>` wouldn't have any effect

I can think of two alternatives here:

Either, we remove the abstraction of combining framing and parsing into a single `decoder` and pass a `framer` and `parser` to the source instead. The benefits would be:
- it would be obvious that the `usize` comes from the frame size, regardless of parsing events out of it
- parsers can stream events without heap allocations

Or, we keep the combined decoder, but use a `SmallVec` so that the `Decoder` has `Item = (SmallVec<Event>, usize)` and the parser exposes `fn parse(&self, bytes: Bytes) -> crate::Result<SmallVec<Event>>;`

Reply:
I implemented the `SmallVec` variant here: 602f301c214a1fe3a28a44f35da33242d1847e59

Reply:
Added to the RFC in 1f016f3da35d9838364f653db641f2766b166b6e